server/src/main/java/org/opensearch/gateway/AsyncShardsFetchPerNode.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.gateway;
+
+import org.opensearch.action.ActionListener;
+import org.opensearch.action.support.nodes.BaseNodeResponse;
+import org.opensearch.action.support.nodes.BaseNodesResponse;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.common.lease.Releasable;
+import org.opensearch.index.shard.ShardId;
+
+import java.util.Map;
+
+/**
+ * This class is responsible for fetching shard data from nodes. It is analogous to AsyncShardFetch class since it fetches
+ * the data in asynchronous manner too.
+ * @param <T>
+ */
+public abstract class AsyncShardsFetchPerNode<T extends BaseNodeResponse> implements Releasable {
+    /**
+     * An action that lists the relevant shard data that needs to be fetched.
+     */
+    public interface Lister<NodesResponse extends BaseNodesResponse<NodeResponse>, NodeResponse extends BaseNodeResponse> {
+        void list(DiscoveryNode[] nodes, Map<ShardId, String> shardsIdMap, ActionListener<NodesResponse> listener);
```

Comment:
The interface looks out of place

Reply:
I am following the structure as per the existing code https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/gateway/AsyncShardFetch.java#L79. More code to be added to this class [AsyncShardsFetchPerNode.java](https://github.com/opensearch-project/OpenSearch/pull/8218/files/1dd48664a5c2b26638d062896d712c9217ffc8d3#diff-1be52136dc0a37d885f47fb54581d5b217565036ef11ddd0a0852a13794a9a1f) while resolving the other sub tasks.